### PR TITLE
Fix dtype for streaming confusion matrix

### PR DESCRIPTION
**Pull request by taion**
Currently, dependents of `_streaming_confusion_matrix` don't actually work correctly with floating-point weights.

### CODE REVIEW

File: tensorflow/python/ops/metrics_impl.py [hunk 1]
```diff
@@ -258,8 +258,11 @@ def _streaming_confusion_matrix(labels, predictions, num_classes, weights=None):
     total_cm: A `Tensor` representing the confusion matrix.
     update_op: An operation that increments the confusion matrix.
   """
+  # TODO: This should be int64 when weights is None, but legacy bug-
+  # compatibility requires keeping this as float64 in that case.
+  cm_dtype = dtypes.float64
```
Comment (taion) [inline tensorflow/python/ops/metrics_impl.py]: The more straightforward fix would be to just invert the branches.

However, if anybody is importing this function from here, and is e.g. dividing this by whatever in Python 2 without doing a float cast, this would then change the numerical results that they see.

Given that this bug has been around for a while, I think it'd perhaps be justifiable to keep the old behavior for cases where weights aren't specified.

I personally wouldn't be affected by a correct fix of just inverting the branches, though, and would be happy to amend this PR if such a fix would be better.

**Comment by martinwicke** [comment]
@poxvoculi your name was all over blame, I was confused too. Probably a merge you did.

@alextp can you take a look?

File: tensorflow/python/kernel_tests/metrics_test.py [hunk 2]
```diff
@@ -3541,7 +3541,7 @@ def testMultipleUpdatesWithWeights(self):
       _enqueue_vector(sess, weights_queue, [1.0])
       _enqueue_vector(sess, weights_queue, [1.0])
       _enqueue_vector(sess, weights_queue, [0.0])
-      _enqueue_vector(sess, weights_queue, [1.0])
```
Comment (alextp) [inline tensorflow/python/kernel_tests/metrics_test.py]: Why this change?

File: tensorflow/python/ops/metrics_impl.py (hunk 1 above)
Comment (alextp) [reply tensorflow/python/ops/metrics_impl.py]: I think inverting the branches is better here.

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/metrics_test.py (hunk 2 above)
Comment (taion) [reply tensorflow/python/kernel_tests/metrics_test.py]: It's the simplest way to exercise the new logic – the old code would fail with this case, since the weight rounds to 0.

File: tensorflow/python/ops/metrics_impl.py (hunk 1 above)
Comment (taion) [reply tensorflow/python/ops/metrics_impl.py]: Sure, will do.

**Comment by taion** [comment]
Updated.

File: tensorflow/python/kernel_tests/metrics_test.py (hunk 2 above)
Comment (taion) [reply tensorflow/python/kernel_tests/metrics_test.py]: Truncate to 0, rather.

**Review by alextp — APPROVED** [review]

**Comment by alextp** [comment]
Jenkins, test this please.

**Comment by alextp** [comment]
Please address the test failure as it is relevant.

**Comment by taion** [comment]
Oops, my bad. I updated the wrong weight for the test. Fixed now.

**Comment by taion** [comment]
Anyway, be aware that if you have any internal consumers of `_streaming_confusion_matrix` without weights in Python 2 that divide by integers (e.g. for building a streaming normalized confusion matrix metric), this would constitute a bit of a breaking change in practice, as in the no-weights case, the confusion matrix tensors will now be of type `int64` rather than `float64` as before.

**Comment by alextp** [comment]
Jenkins, test this please.

**Comment by alextp** [comment]
Jenkins, test this please.

File: tensorflow/python/ops/metrics_impl.py [hunk 3]
```diff
@@ -259,7 +259,7 @@ def _streaming_confusion_matrix(labels, predictions, num_classes, weights=None):
     update_op: An operation that increments the confusion matrix.
   """
   # Local variable to accumulate the predictions in the confusion matrix.
-  cm_dtype = dtypes.int64 if weights is not None else dtypes.float64
+  cm_dtype = dtypes.float64 if weights is not None else dtypes.int64
```
Comment (martinwicke) [inline tensorflow/python/ops/metrics_impl.py]: Can you make this float64 unconditionally? That is a slightly less breaking change, since it doesn't change the return type in the case that actually works (without weights). Changing the type in the case that was broken is a plain bugfix.

**Review by martinwicke — CHANGES_REQUESTED** [review]

**Comment by martinwicke** [comment]
One change request from API review.

File: tensorflow/python/ops/metrics_impl.py (hunk 3 above)
Comment (taion) [reply tensorflow/python/ops/metrics_impl.py]: This was the initial implementation per https://github.com/tensorflow/tensorflow/pull/12714/commits/4b521d0992fbdd7d9f18a708111e2501689f7994. @alextp explicitly asked for it to be changed to the current implementation per https://github.com/tensorflow/tensorflow/pull/12714/commits/4b521d0992fbdd7d9f18a708111e2501689f7994#r137026663.

I'd be happy to change it back to my original implementation (with the FIXME?), but please confirm that this is as desired.

In some moral sense, `int64` probably is more correct with no weights (or with integer weights), but I share your concerns about API breakage. Just want to confirm this given the specific earlier request to use `int64` in the no-weights case.

Comment (taion) [reply tensorflow/python/ops/metrics_impl.py]: There are no semver implications here per se, since `_streaming_confusion_matrix` is not part of the public API, and in fact no public APIs change.

Comment (martinwicke) [reply tensorflow/python/ops/metrics_impl.py]: In every place we use it, we convert to float. So it may make more sense to always have it in float so we don't have to do that. 

But you're right, technically it's an implementation detail and nobody would need to ever know.

Comment (taion) [reply tensorflow/python/ops/metrics_impl.py]: Yeah. I'm fine either way though. Just let me know which you want.

Comment (martinwicke) [reply tensorflow/python/ops/metrics_impl.py]: Let's go with all float.

Comment (taion) [reply tensorflow/python/ops/metrics_impl.py]: Done.

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Jenkins, test this please.
